File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        vector<int> dp(n + 1, 0);
        int oneStep, twoStep;
        for(int i = 2; i <= n; i++){
            oneStep = cost[i - 1] + dp[i - 1];
            twoStep = cost[i - 2] + dp[i - 2];
            dp[i] = min(oneStep, twoStep);
        }
        return dp[n];
    }
};

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp = dict()
        def dfs(i: int) -> int:
            if i >= len(cost):
                return 0
            if i in dp:
                return dp[i]
            dp[i] = min(dfs(i + 1), dfs(i + 2)) + cost[i]
            return dp[i]
        return min(dfs(0), dfs(1))